content/posts/2021/reana-client-0.7.3.md

```diff
@@ -0,0 +1,346 @@
+-        environment: 'python:3.9-slim'
++        environment: 'foo:bar'
+         commands:
+           - python "${script_path}" "${sleeptime}"
+```
+
+Let us call the validator, passing the environments option:
+
+```console
+$ reana-client validate -f reana.yaml --environments
+==> Verifying REANA specification file... my-analysis/reana.yaml
+  -> SUCCESS: Valid REANA specification file.
+==> Verifying workflow parameters and commands...
+==> Verifying environments in REANA specification file...
+  -> SUCCESS: Environment image foo:bar has the correct format.
+  -> WARNING: Environment image foo:bar does not exist locally.
+  -> WARNING: Environment image foo:bar does not exist in Docker Hub: "Resource not found"
+  -> ERROR: Environment image foo:bar does not exist locally or remotely.
+```
+
+Let us analyse the environment validation output. First, we get a
+success message informing us that the image has the correct format.
+This is because, as expected, it has the image name and the tag, so
+in that regard it is correct. Then we get two warnings telling us
+that the image was not found either locally nor in the Docker Hub
+registry.
+As a consequence of this, we can see the last error message, telling
+us that the image does not exist, thus making the validation fail and
+stop its execution.
+
+Let us change the environment to use a valid one. We are going to
+revert back to the `python` image but using the `latest` tag instead:
+
+```diff
+   specification:
+     steps:
+       - name: run-script
+-        environment: 'foo:bar'
++        environment: 'python:latest'
+         commands:
+           - python "${script_path}" "${sleeptime}"
+```
+
+This is something that we strongly discourage as the `latest` tag can
+evolve in time, i.e. the image that is pointing today might, and
+probably will be different in the future. Let us run the validator
+again:
+
+```console
+$ reana-client validate -f reana.yaml --environments
+==> Verifying REANA specification file... my-analysis/reana.yaml
+  -> SUCCESS: Valid REANA specification file.
+==> Verifying workflow parameters and commands...
+==> Verifying environments in REANA specification file...
+  -> WARNING: Using 'latest' tag is not recommended in python environment image.
+  -> WARNING: Environment image python:latest does not exist locally.
+  -> SUCCESS: Environment image python:latest exists in Docker Hub.
+  -> WARNING: UID/GIDs validation skipped, specify `--pull` to enable it.
+```
+
+As part of the environment verification, we get a message warning us
+about the usage of the `latest` tag. Additionally, we have two checks to
+verify if the image exists locally and remotely. In this particular
+case, since we do not have the `python:latest` image pulled locally,
+a warning is displayed. On the other hand, the `python:latest` image
+exists in Docker Hub registry, this we get a success message.
+
+
+The last warning displayed is related to the user ID (UID) and
```

Comment:
Consider new section about UID and GID, see above.

Reply:
If we go for the new section, we can start philosophically like:

The reproducible workflows should ideally not depend on the given user who executes them. The workflow should give the same result regardless of user identity, such as  UID (user ID) and GID (group ID) known from Unix systems.
For security reasons, REANA executes workflows as UID=1000 (see here), and expects GID=0 to be able to share filesystem write rights across multiple nodes (see here). This technicalities usually don't matter. However, in certain cases, you may need to execute your workflow under a different user identity, for example because the environment image expects certain user privileges. REANA allows this by setting `kubernetes_uid` workflow hint (see here). However, it can happen that workflow image user ID expectations are conflicting with the declared UID, which leads to conflicts at the workflow execution time.

REANA 0.7.3 validation of environments allows to catch these situations early by inspecting the workflow specification and the container images. In order  to be able to do such inspection, you must have a running `docker` on the machine were you run `reana-client`. The environment image has to be pulled locally, which also consumes disk space. The validation of image UID and GID is therefore triggered by another command line option called `--pull`.  Please use this option only if you are used to working with `docker` images locally.

Let us illustrate how the enviroment image UID and GID validation works by rerunning our past example:

...

(and then you go on)

Reply:
I think we don't have any reference for `kubernetes_uid` to link the "see here".